File: java/BurstBalloons.java
class Solution {
    public int maxCoins(int[] nums) {
        
        int length = nums.length;
        int[] newNums = new int[length + 2];
        newNums[0] = 1;
        newNums[length + 1] = 1;
        for (int i = 0; i < length; i++){
            newNums[i + 1] = nums[i];
        }
                        
        int[][] res = new int[length + 2][length + 2];
        
        for (int k = 2; k < length + 2; k++){
            for (int left = 0; left < length + 2 - k; left++){
                int right = k + left;
                for (int i = left + 1; i < right; i++){
                    res[left][right] = Math.max(res[left][right], newNums[i] * newNums[left] * newNums[right] + res[left][i] + res[i][right]);
                }
            }
        }
        // (0, length + 1), does not include 0 and length + 1
        return res[0][length + 1];
    }
}

File: python/BurstBalloons.py
class Solution(object):
    def maxCoins(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        newNums = [1] + [num for num in nums] + [1]
        n = len(newNums)
        dp = [[0] * n for _ in range(n)]
        
        for k in range(2, n):
            for left in range(0, n - k):
                right = left + k
                for i in range(left + 1, right):
                    dp[left][right] = max(dp[left][right], newNums[i] * newNums[left] * newNums[right] + dp[left][i] + dp[i][right])            
    
        return dp[0][n - 1]
